doc/try_into.md

```diff
@@ -0,0 +1,132 @@
+% What #[derive(TryInto)] generates
+
+This derive allows you to convert enum variants into their corresponding
+variant types.
+One thing to note is that this derive doesn't actually generate an
+implementation for the `TryInto` trait.
+Instead it derives `TryFrom` for each variant in the enum and thus has an
+indirect implementation of `TryInto` as recommended by the
+[docs](https://doc.rust-lang.org/core/convert/trait.TryInto.html).
+
+# Structs
+
+Deriving `TryInto` for structs is not supported because there is no failing
+mode. Use `#[derive(Into)]` instead. `TryInto` will automatically get a
+blanket implementation through `TryFrom`, automatically derived from `From`,
+which `#[derive(Into)]` produces.
+
+# Enums
+
+When deriving `TryInto` for an enum, each enum variant gets its own
+`TryFrom` implementation.
+For instance, when deriving `TryInto` for an enum link this:
+
+```rust
+#![feature(try_from)]
+# #[macro_use] extern crate derive_more;
+# fn main(){}
+#[derive(TryInto)]
+enum MixedInts {
+    SmallInt(i32),
+    BigInt(i64),
+    TwoSmallInts(i32, i32),
+    NamedSmallInts { x: i64, y: i64 },
+    UnsignedOne(u32),
+    UnsignedTwo(u32),
+}
+```
+
+Code like this will be generated:
+
+```rust
+# #![feature(try_from)]
+# enum MixedInts {
+#     SmallInt(i32),
+#     BigInt(i64),
+#     TwoSmallInts(i32, i32),
+#     NamedSmallInts { x: i64, y: i64 },
+#     UnsignedOne(u32),
+#     UnsignedTwo(u32),
+# }
+impl ::std::convert::TryFrom<MixedInts> for (i32) {
+    type Error = ();
+    fn try_from(value: MixedInts) -> Result<Self, Self::Error> {
+        match value {
+            MixedInts::SmallInt(__0) => Ok(__0),
+            _ => Err(()),
+        }
+    }
+}
+impl ::std::convert::TryFrom<MixedInts> for (i64) {
+    type Error = ();
+    fn try_from(value: MixedInts) -> Result<Self, Self::Error> {
+        match value {
+            MixedInts::BigInt(__0) => Ok(__0),
+            _ => Err(()),
+        }
+    }
+}
+impl ::std::convert::TryFrom<MixedInts> for (i32, i32) {
+    type Error = ();
+    fn try_from(value: MixedInts) -> Result<Self, Self::Error> {
+        match value {
+            MixedInts::TwoSmallInts(__0, __1) => Ok((__0, __1)),
+            _ => Err(()),
+        }
+    }
+}
+impl ::std::convert::TryFrom<MixedInts> for (i64, i64) {
+    type Error = ();
+    fn try_from(value: MixedInts) -> Result<Self, Self::Error> {
+        match value {
+            MixedInts::NamedSmallInts { x: __0, y: __1 } => Ok((__0, __1)),
+            _ => Err(()),
+        }
+    }
+}
+```
+
+When multiple variants have the same type, no `TryInto` is derived for them.
```

Comment:
Maybe I'm missing something, but I think it actually makes sense to derive `TryInto` for variants with the same type as well. Because both can be converted into their inner type. This wasn't the case for `From` because the other way around you don't know which enum variant to choose.

Reply:
In that example, there would be two `TryFrom<MixedInts> for u32`, that wouldn't work out.

Reply:
You could have one `TryFrom<MixedInts> for u32` where multiple match arms return an `Ok(....)` right?

Reply:
Oh, yes, indeed.